apps/test-aeat-6012/src/main.cpp

```diff
@@ -0,0 +1,113 @@
+#include "AEAT6012.h"
+#include "mbed.h"
+
+constexpr auto ENCODER_READ_PERIOD = 1ms;
+constexpr auto PRINTF_PERIOD       = 1000ms;
+
+enum Test_Mode_E { TEST_BLOCKING, TEST_ASYNC };
+
+void encoder_read_blocking(void);
+void encoder_read_async(void);
+void callback(void);
+
+Thread thread;
+Timer timer;
+
+volatile bool encoder_position_updated         = false;
+volatile float encoder_angle_deg               = 0;
+volatile float encoder_angular_vel_deg_per_sec = 0;
+volatile uint32_t encoder_pos_read_time_us     = 0;
+volatile uint32_t encoder_vel_read_time_us     = 0;
+
+Encoder::AEAT6012 encoder(SPI_CS, SPI_MISO, SPI_SCK);
+
+// Modify this
+Test_Mode_E test_mode = TEST_BLOCKING;
+
+int main() {
+  switch (test_mode) {
+    case TEST_BLOCKING:
+      thread.start(encoder_read_blocking);
+      break;
+
+    case TEST_ASYNC:
+      thread.start(encoder_read_async);
+      break;
+
+    default:
+      break;
+  }
+
+  while (true) {
+    printf("Encoder position degrees: %.3f\r\n", encoder_angle_deg);
+    printf("Read time: %lu us\r\n\r\n", encoder_pos_read_time_us);
+
+    printf("Encoder angular velocity deg/s: %.3f\r\n", encoder_angular_vel_deg_per_sec);
+    printf("Read time: %lu us\r\n\r\n", encoder_vel_read_time_us);
+
+    ThisThread::sleep_for(PRINTF_PERIOD);
+  }
+}
+
+void encoder_read_blocking(void) {
+  printf("\r\n--- AEAT-6012 Blocking Driver Test ---\r\n\r\n");
+  float measurement;
+  encoder.reset();
+
+  while (true) {
+    timer.reset();
+    timer.start();
+
+    while (!encoder.getAngleDeg(measurement))
+      ;
+
+    timer.stop();
+    encoder_angle_deg        = measurement;
+    encoder_pos_read_time_us = std::chrono::duration_cast<std::chrono::microseconds>(timer.elapsed_time()).count();
+
+    ThisThread::sleep_for(1ms);
+
+    timer.reset();
+    timer.start();
+
+    while (!encoder.getAngularVelocityDegPerSec(measurement))
+      ;
+
+    timer.stop();
```

Comment:
does this read until an error happens? if so why?

Reply:
it reads until an error doesn't happen

Reply:
oh wait how often does the read fail (%wise)?  maybe do like 1000 successive reads and count successes and failures
It should be failing almost never.

Also, you test driver shuoldnt take this approach. it should just read once and that should work. It more closely resemebles how users will use ur code

Reply:
lol it never fails cuz read() always returns true

Reply:
That's what it should be returning right?

Reply:
i thought that if the SPI transfer failed for some reason or something it would return false. ig if never fails thats good?